client/components/Registration/Register.jsx

```diff
@@ -1,94 +1,112 @@
-import React, { useState } from 'react'
+import React from 'react'
+// import { useFormik } from 'formik'
+import { useFormik } from 'formik'
 import { useHistory } from 'react-router-dom'
 import { registerUser } from './registerHelper'
 import { useAuth0 } from '@auth0/auth0-react'
+import * as Yup from 'yup'
 
-export function Register () {
+const registerSchema = Yup.object().shape({
+  firstName: Yup.string()
+    .min(2, 'Too Short!')
+    .max(15, 'Too Long!')
+    .required('Required'),
+  lastName: Yup.string()
+    .required('Required')
+    .min(2, 'Too Short!')
+    .max(15, 'Too Long!'),
+  username: Yup.string()
+    .min(2, 'Too Short!')
+    .max(15, 'Too Long!')
+    .required('Required'),
+  gardenId: Yup.number()
+    .required('Required')
+})
+
+export default function Register () {
   const authUser = useAuth0().user
+  const history = useHistory()
 
-  const [form, setForm] = useState({
-    firstName: '',
-    lastName: '',
-    username: '',
-    gardenId: null
+  const formik = useFormik({
+    initialValues: {
+      firstName: '',
+      lastName: '',
+      username: '',
+      gardenId: null
+    },
+    onSubmit: values => {
+      console.log('hello')
+      registerUser(values, authUser, history.push)
+    },
+    validationSchema: registerSchema
   })
-  const history = useHistory()
 
-  function handleChange (e) {
-    const { name, value } = e.target
-    setForm({
-      ...form,
-      [name]: value
-    })
-  }
+  return (
+    <>
+      <section className='flex-container'>
+        <form className='column-6' onSubmit={formik.handleSubmit}>
+          <div className="field">
+            <label htmlFor='firstName' className='form-label'>First Name</label>
+            {formik.errors.firstName && formik.touched.firstName
+              ? (<p className="inputError">{formik.errors.firstName}</p>)
+              : null}
```

Comment:
This approach to showing error messages can be packaged up into a function to remove the duplication for each input. For example, with a function like:

```
function showAnyErrors (inputName) {
  return formik.errors[inputName] && formik.touched[inputName]
  ? <p className='inputError'>{formik.errors[inputName]}</p>
  : null
}
```

you could change this section to: `{showAnyErrors('firstName')}`

Let me know if this doesn't make sense and I can add some clarity.

Reply:
Ooosh love this